File: State Photos/state_photos/framing.py
"""Shared math for how a cover image is placed inside a state shape (map + preview)."""

from __future__ import annotations


def clamp_focus(v: float) -> float:
    return max(0.0, min(100.0, float(v)))


def clamp_zoom(v: float) -> float:
    return max(1.0, min(2.5, float(v)))


def photo_framing_values(photo) -> tuple[float, float, float]:
    """Read framing from a Photo model (handles legacy NULLs)."""
    fx = float(photo.focus_x) if photo.focus_x is not None else 50.0
    fy = float(photo.focus_y) if photo.focus_y is not None else 50.0
    z = float(photo.frame_zoom) if photo.frame_zoom is not None else 1.0
    return clamp_focus(fx), clamp_focus(fy), clamp_zoom(z)


def pattern_image_transform(focus_x: float, focus_y: float, zoom: float) -> str:
    """
    Transform for the <image> inside a pattern with patternContentUnits=objectBoundingBox,
    width/height 1, preserveAspectRatio xMidYMid slice. focus_* are 0–100 like CSS object-position;
    50,50 is centered. zoom 1 = default fill; higher = zoom in (tighter crop).
    """
    fx = clamp_focus(focus_x)
    fy = clamp_focus(focus_y)
    z = clamp_zoom(zoom)
    # Pan range tuned so sliders feel useful on irregular state shapes
    k = 0.35
    dx = (50.0 - fx) / 100.0 * k
    dy = (50.0 - fy) / 100.0 * k
    return f"translate(0.5 0.5) scale({z}) translate(-0.5 -0.5) translate({dx:.5f} {dy:.5f})"

File: lib/statePhotos/framing.ts
export function clampFocus(v: number): number {
  return Math.max(0, Math.min(100, Number(v) || 0));
}

export function clampZoom(v: number): number {
  return Math.max(1, Math.min(2.5, Number(v) || 1));
}

export function photoFramingValues(row: {
  focus_x: number | null;
  focus_y: number | null;
  frame_zoom: number | null;
}): [number, number, number] {
  const fx = row.focus_x != null ? clampFocus(row.focus_x) : 50;
  const fy = row.focus_y != null ? clampFocus(row.focus_y) : 50;
  const z = row.frame_zoom != null ? clampZoom(row.frame_zoom) : 1;
  return [fx, fy, z];
}

export function patternImageTransform(focusX: number, focusY: number, zoom: number): string {
  const fx = clampFocus(focusX);
  const fy = clampFocus(focusY);
  const z = clampZoom(zoom);
  const k = 0.35;
  const dx = ((50 - fx) / 100) * k;
  const dy = ((50 - fy) / 100) * k;
  return `translate(0.5 0.5) scale(${z}) translate(-0.5 -0.5) translate(${dx.toFixed(5)} ${dy.toFixed(5)})`;
}
